File: backend/Testing/Browser/Page/External/MailPages.cs
using Microsoft.Playwright;
using Testing.Services;

namespace Testing.Browser.Page.External;

public abstract class MailInboxPage : BasePage<MailInboxPage>
{
    public string MailboxId { get; private set; }
    public ILocator EmailLocator { get; }

    public static MailInboxPage Get(IPage page, string mailboxId)
    {
        return TestingEnvironmentVariables.IsDev
            ? new MailDevInboxPage(page, mailboxId)
            : new MailinatorInboxPage(page, mailboxId);
    }

    public MailInboxPage(IPage page, string url, ILocator testLocator, string mailboxId, ILocator emailLocator)
    : base(page, url, testLocator)
    {
        MailboxId = mailboxId;
        EmailLocator = emailLocator;
    }

    protected abstract MailEmailPage GetEmailPage();

    public async Task<MailInboxPage> GotoMailbox(string mailboxId)
    {
        MailboxId = mailboxId;
        return await Goto();
    }

    public async Task<MailEmailPage> OpenEmail(int index = 0)
    {
        await EmailLocator.Nth(index).ClickAsync();
        return await GetEmailPage().WaitFor();
    }
}

public abstract class MailEmailPage : BasePage<MailEmailPage>
{
    protected readonly ILocator bodyLocator;

    public ILocator ResetPasswordButton => bodyLocator.GetByRole(AriaRole.Link, new() { Name = "Reset password" });

    public MailEmailPage(IPage page, string? url, ILocator bodyLocator) : base(page, url, bodyLocator)
    {
        this.bodyLocator = bodyLocator;
    }

    public Task ClickVerifyEmail()
    {
        return bodyLocator.GetByRole(AriaRole.Link, new() { Name = "Verify e-mail" }).ClickAsync();
    }

    public Task ClickResetPassword()
    {
        return ResetPasswordButton.ClickAsync();
    }

    public virtual Task<string> GetFirstLanguageDepotUrl()
    {
        return bodyLocator.Locator($"a[href*='{TestingEnvironmentVariables.ServerBaseUrl}']").First.GetAttributeAsync("href");
    }
}

File: frontend/tests/pages/mailPages.ts
import { type Locator, type Page, expect } from '@playwright/test';
import { BasePage } from './basePage';
import { serverBaseUrl } from '../envVars';

export enum EmailSubjects {
  VerifyEmail = 'Verify your e-mail address',
  ForgotPassword = 'Forgot your password?',
  PasswordChanged = 'Your password was changed',
  ProjectInvitation = 'Project invitation:',
}

export abstract class MailInboxPage extends BasePage {
  public mailboxId: string;
  readonly emailLocator: Locator;

  constructor(page: Page, testLocator: Locator, url: string, mailboxId: string, emailLocator: Locator) {
    super(page, testLocator, url);
    this.mailboxId = mailboxId;
    this.emailLocator = emailLocator;
  }

  abstract getEmailPage(): MailEmailPage;
  abstract refreshEmails(): Promise<void>;

  async gotoMailbox(mailboxId: string): Promise<MailInboxPage> {
    this.mailboxId = mailboxId;
    return await this.goto();
  }

  async openEmail(subject: EmailSubjects, index = 0): Promise<MailEmailPage> {
    const email = this.emailLocator.locator(`:text("${subject}")`).nth(index);
    // Emails may not be immediately available, so if they aren't, refresh the email list until they show up
    await expect(async () => {
      if (!await email.isVisible()) {
        await this.refreshEmails();
      }
      await email.click();
    }, `Failed to find email: ${subject} (${index})`).toPass({timeout: 10_000}); // This auto-retries on a reasonable schedule
    return await this.getEmailPage().waitFor();
  }
}

export abstract class MailEmailPage extends BasePage {
  readonly bodyLocator: Locator;
  public get resetPasswordButton(): Locator { return this.bodyLocator.getByRole('link', {name: 'Reset password'}); }

  constructor(page: Page, bodyLocator: Locator, url?: string) {
    super(page, bodyLocator, url);
    this.bodyLocator = bodyLocator;
  }

  clickVerifyEmail(): Promise<void> {
    return this.bodyLocator.getByRole('link', {name: 'Verify e-mail'}).click();
  }

  clickResetPassword(): Promise<void> {
    return this.resetPasswordButton.click();
  }

  getFirstLanguageDepotUrl(): Promise<string | null> {
    return this.bodyLocator.locator(`a[href*='${serverBaseUrl}']`).first().getAttribute('href');
  }

  clickFirstLanguageDepotUrl(): Promise<void> {
    return this.bodyLocator.locator(`a[href*='${serverBaseUrl}']`).first().click();
  }
}
